helix-core/src/main/java/org/apache/helix/manager/zk/ZKHelixManager.java

```diff
@@ -225,9 +226,24 @@ public ZKHelixManager(String clusterName, String instanceName, InstanceType inst
       String zkAddress, HelixManagerStateListener stateListener,
       HelixManagerProperty helixManagerProperty) {
```

Comment:
helixManagerProperty.RealmAwareZkConnectionConfig now contains zk connection information that possibly conflicts with zkAddress. This is dangerous.
Actually, I already see one potential bug in the method getMetadataStoreConnectionString. Which always returns the zkAddress only without checking the connectionConfig.

Please ensure we are only recording realmAwareZkConnectionConfig. Let's try to remove _zkAddress for good. Or at least update this field according to the realmAwareZkConnectionConfig input.

Other configurations such as sessionTimeout are in the same situation.

Reply:
The documentation makes it clear that zkAddress given will be overriden if a valid ZkConnectionConfig is supplied.

We cannot remove zkAddress as that would break backward-compatibility.

Reply:
Ideally, we should deprecate this one but create a constructor with clean parameters.

Besides a real concern is that if one passes a wrong or invalid zkAddress with this constructor (since he or she assumes it will be overridden by ZkConnectionConfig), then getMetadataStoreConnectionString will return this invalid value too. And the WAGED rebalancer won't work.
A slightly less serious issue would be, most of our logs are using zkAddress, so it would be harder for us to debug.

Reply:
@jiajunwang Then we could play it safe by adding a check so that we do not allow instantiation of a ZkHelixManager instance when both zkAddress and ZkConnectionConfig are set. How does that sound?